psc-flink/src/main/java/com/pinterest/flink/streaming/connectors/psc/table/PscDynamicSource.java

```diff
@@ -309,11 +357,41 @@ public DataStream<RowData> produceDataStream(
                         execEnv.fromSource(
                                 pscSource, watermarkStrategy, "PscSource-" + tableIdentifier);
                 
-                // Let Flink handle source parallelism automatically (defaults to partition count)
-                // Only add rescale if explicitly enabled to redistribute data across downstream operators
+                // Source parallelism is determined by partition count (Flink's default for Kafka-like sources)
+                // We do NOT set it explicitly even if scan.parallelism is configured, because:
+                // - A Kafka source can only have as many active subtasks as there are partitions
+                // - Setting higher parallelism would create idle subtasks
+                // - Instead, we use rescale() to redistribute data to the intended downstream parallelism
+                
                 DataStream<RowData> resultStream = sourceStream;
+                
+                // Determine the intended downstream parallelism for rate limiting
+                // This is scan.parallelism if set, otherwise global default parallelism
+                int intendedParallelism = getIntendedParallelism(execEnv);
+                
+                // Apply rescale FIRST if enabled
+                // This redistributes data from source parallelism (= partition count) to intended parallelism
+                // Ensures all downstream subtasks (including rate limiters) receive traffic
                 if (enableRescale) {
-                    resultStream = sourceStream.rescale();
+                    resultStream = resultStream.rescale();
+                }
+                
+                // Apply rate limiting AFTER rescale if configured
+                // Rate limiter parallelism must match the actual parallelism of incoming data:
+                // - If rescale enabled: use intendedParallelism (all subtasks are active after rescale)
+                // - If rescale disabled: use source parallelism (rate limiter stays with source)
+                if (isRateLimitingEnabled(rateLimitRecordsPerSecond)) {
+                    int rateLimiterParallelism = getRateLimiterParallelism(
```

Comment:
nit: It seems straightforward as `enableRescale ? intendedParallelism : sourceStream.getParallelism()`